src/Bicep.Core/Parsing/Parser.cs

```diff
@@ -1,370 +1,184 @@
-            var diagnosticWriter = ToListDiagnosticWriter.Create();
-            var lexer = new Lexer(new SlidingTextWindow(text), diagnosticWriter);
-            lexer.Lex();
-
-            this.lexerDiagnostics = diagnosticWriter.GetDiagnostics().ToImmutableArray();
-
-            this.reader = new TokenReader(lexer.GetTokens());
-        }
-
-        public ProgramSyntax Program()
-        {
+        public Parser(string text)
+        {
+            // treating the lexer as an implementation detail of the parser
+            var diagnosticWriter = ToListDiagnosticWriter.Create();
+            var lexer = new Lexer(new SlidingTextWindow(text), diagnosticWriter);
+            lexer.Lex();
+
+            this.lexerDiagnostics = diagnosticWriter.GetDiagnostics().ToImmutableArray();
+
+            this.reader = new TokenReader(lexer.GetTokens());
+        }
+
+        public ProgramSyntax Program()
+        {
             var declarationsOrTokens = new List<SyntaxBase>();
 
-            while (!this.IsAtEnd())
-            {
-                // this produces either a declaration node, skipped tokens node or just a token
-                var declarationOrToken = Declaration();
-                declarationsOrTokens.Add(declarationOrToken);
-
-                // if skipped node is returned above, the newline is not consumed
-                // if newline token is returned, we must not expect another (could be a beginning of a declaration)
-                if (declarationOrToken is IDeclarationSyntax)
-                {
-                    // declarations must be followed by a newline or the file must end
-                    var newLine = this.WithRecoveryNullable(this.NewLineOrEof, RecoveryFlags.ConsumeTerminator, TokenType.NewLine);
-                    if (newLine != null)
+            while (!this.IsAtEnd())
+            {
+                // this produces either a declaration node, skipped tokens node or just a token
+                var declarationOrToken = Declaration();
+                declarationsOrTokens.Add(declarationOrToken);
+
+                // if skipped node is returned above, the newline is not consumed
+                // if newline token is returned, we must not expect another (could be a beginning of a declaration)
+                if (declarationOrToken is IDeclarationSyntax)
+                {
+                    // declarations must be followed by a newline or the file must end
+                    var newLine = this.WithRecoveryNullable(this.NewLineOrEof, RecoveryFlags.ConsumeTerminator, TokenType.NewLine);
+                    if (newLine != null)
+                    {
+                        declarationsOrTokens.Add(newLine);
+                    }
+                }
+            }
+
+            var endOfFile = reader.Read();
+
+            return new ProgramSyntax(declarationsOrTokens, endOfFile, this.lexerDiagnostics);
+        }
+
+        public SyntaxBase Declaration() =>
```

Comment:
>Declaration [](start = 26, length = 11)

What happens if the file has one or more decorators but no declaration after it/them?

Reply:
The behavior should be that you get syntax coloring and validation within the annotation (as well as hovers and such) and errors due to the missing declaration.

Reply:
Yeah definitely something I missed. It can lead to incomplete AST if the dangling decorators are consumed by the parser. I'll update the code to handle it.

Reply:
Borrowed ideas from Roslyn and TypeScript and `MissingDeclarationSyntax` to cover this.

Reply:
Makes sense.